### DISCUSSION TITLE
About destroy and memory relationship

### DISCUSSION THREAD

**Discussion by redcamel** [gpuweb/gpuweb#3538] [Q&A]
Just by destroying the gpudevice
Doesn't it free the associated memory?

Do I have to manage all the buffers or textures created with the gpuDevice at the same time?

**Comment by Kangz** [comment]
Yes, calling `GPUDevice.destroy` is meant to free all the GPU memory used by the device, in a safe manner. This means that you don't need to wait for the textures and buffers to be finished being used before destroying the device. The resources will be implicitly destroyed as well when the device is destroyed.

**Reply by redcamel** [reply]
for example
When calling multiple WebGPU Contexts at once...
At some point, the WebGPU Context cannot be called.
I want to explicitly remove existing Contexts or situations.

When writing test cases, you need to create a lot of WebGPU Contexts.
Executing sequentially... even if you explicitly destroy
I had the same problem as above and asked.

```
code : 0

message
D3D12 create command queue failed with E_OUTOFMEMORY (0x8007000E)
    at CheckHRESULTImpl (..\..\third_party\dawn\src\dawn\native\d3d12\D3D12Error.cpp:74)
    at Initialize (..\..\third_party\dawn\src\dawn\native\d3d12\DeviceD3D12.cpp:87)
    at Create (..\..\third_party\dawn\src\dawn\native\d3d12\DeviceD3D12.cpp:70)
    
name : OperationError
```

**Reply by redcamel** [reply]
The same goes for using the unconfigure() method of the WebGPU Context together. Can you give me an answer?

**Reply by Kangz** [reply]
> When calling multiple WebGPU Contexts at once...

I'm not sure what you mean. Is it while using multiple `GPUDevices` at once? That should work, and not run out of devices resources too fast. Do you have a repro case you could share?

**Reply by redcamel** [reply]
make webGPU context => destroy
make webGPU context => destroy
make webGPU context => destroy
make webGPU context => destroy
make webGPU context => destroy
make webGPU context => destroy
make webGPU context => destroy
make webGPU context => destroy
make webGPU context => destroy
.......
If you repeat it, at some point you can no longer create it.

In the case of my PC, it can no longer be created after about 64 runs.

**Reply by redcamel** [reply]
@Kangz 

simple test URL - https://redcamel.github.io/Rnd_Doc/host/destroyTest.html
![image](https://user-images.githubusercontent.com/820587/197455423-92f6fb67-ca16-47ed-a7f2-f95b025f2297.png)
After running it in Canary, please check the console window.

**Reply by redcamel** [reply]
It feels like something is different depending on the individual graphics card.
On my computer, it stops at about 64.
On my personal computer, it performs just fine.

**Reply by rayxc** [reply]
> Yes, calling `GPUDevice.destroy` is meant to free all the GPU memory used by the device, in a safe manner. This means that you don't need to wait for the textures and buffers to be finished being used before destroying the device. The resources will be implicitly destroyed as well when the device is destroyed.

but the gpudevice is still useful，only some gputexture is not used，how to release the resource used by the gputexture? I has use GPUTexture.destroy，but the video memory was not released

**Reply by kainino0x** [reply]
> but the gpudevice is still useful，only some gputexture is not used，how to release the resource used by the gputexture? I has use GPUTexture.destroy，but the video memory was not released

It is supposed to be. Please file a browser bug with the appropriate browser, and include a minimized reproduction case that we can test with.

**Reply by greggman** [reply]
That example seems like it's keeping references. Each `console.log(initInfo)` puts references into the log. Even though the code called `device.destroy` the canvas elements are sticking around and still referenceable from the log. Not sure removing that will fix things but just pointing out the sample as accumulating objects.

![Screenshot 2025-06-05 at 14 30 05](https://github.com/user-attachments/assets/b3fc8a43-da59-4069-bb8a-cf91a1cb071c)

**Comment by Frendysemon** [comment]
Yep, destroying the gpudevice should free the resources. I still like calling destroy on big buffers and textures when I'm done with them.
